File: 49-GroupAnagrams.java
/*

49. Group Anagrams
Medium

Given an array of strings, group anagrams together.

Example:

Input: ["eat", "tea", "tan", "ate", "nat", "bat"],
Output:
[
  ["ate","eat","tea"],
  ["nat","tan"],
  ["bat"]
]

Note:

    All inputs will be in lowercase.
    The order of your output does not matter.

Link :

*/

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String,List<String>> lookupMap = new HashMap<>();
        List<List<String>> resultList = new ArrayList<>();
        for (String str:strs){
            char tempArray[] = str.toCharArray();
            Arrays.sort(tempArray);
            if (lookupMap.containsKey(new String(tempArray))){
                lookupMap.get(new String(tempArray)).add(str);
            }
            else{
                List<String> tempList = new ArrayList<>();
                tempList.add(str);
                lookupMap.put(new String(tempArray),tempList);
            }
        }
        for(Map.Entry<String,List<String>> entry : lookupMap.entrySet()){
                resultList.add(entry.getValue());
            }
        return resultList;
    }
}

File: 49-GroupAnagrams/49-GroupAnagrams.py
"""

49. Group Anagrams

Given an array of strings strs, group the

together. You can return the answer in any order.

Example 1:

Input: strs = ["eat","tea","tan","ate","nat","bat"]

Output: [["bat"],["nat","tan"],["ate","eat","tea"]]

Explanation:

    There is no string in strs that can be rearranged to form "bat".
    The strings "nat" and "tan" are anagrams as they can be rearranged to form each other.
    The strings "ate", "eat", and "tea" are anagrams as they can be rearranged to form each other.

Example 2:

Input: strs = [""]

Output: [[""]]

Example 3:

Input: strs = ["a"]

Output: [["a"]]

Constraints:

    1 <= strs.length <= 104
    0 <= strs[i].length <= 100
    strs[i] consists of lowercase English letters.

"""

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        mapp = defaultdict(list)
        for s in strs:
            ss = sorted(s, key = lambda x : ord(x))
            mapp["".join(ss)].append(s)
        return list(mapp.values())
